partiql-eval/src/pattern_match.rs

```diff
@@ -0,0 +1,150 @@
+        match (ch, is_escaped) {
+            (_, false) if Some(ch) == escape_ch => escaped = true,
+            ('%', false) => {
+                if !is_any {
+                    buf.push_str(".*?")
+                }
+                wildcard = true;
+            }
+            ('_', false) => buf.push('.'),
+            _ => {
+                if is_meta_character(ch) {
+                    buf.push('\\'); // regex-escape the next character
+                }
+                buf.push(ch);
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use regex::Regex;
+    use std::collections::{BTreeSet, HashSet};
+
+    #[test]
+    fn like() {
+        assert_eq!(like_to_re_pattern("foo", Some('\\')), r#"^foo$"#);
+        assert_eq!(like_to_re_pattern("%foo", Some('\\')), r#"^.*?foo$"#);
+        assert_eq!(like_to_re_pattern("foo%", Some('\\')), r#"^foo.*?$"#);
+        assert_eq!(like_to_re_pattern("foo%bar", Some('\\')), r#"^foo.*?bar$"#);
+        assert_eq!(like_to_re_pattern("foo%%bar", Some('\\')), r#"^foo.*?bar$"#);
+        assert_eq!(
+            like_to_re_pattern("foo%%%bar", Some('\\')),
+            r#"^foo.*?bar$"#
+        );
+        assert_eq!(
+            like_to_re_pattern("foo%%%%bar", Some('\\')),
+            r#"^foo.*?bar$"#
+        );
+        assert_eq!(
+            like_to_re_pattern("%foo%%%%bar%", Some('\\')),
+            r#"^.*?foo.*?bar.*?$"#
+        );
+        assert_eq!(
+            like_to_re_pattern("%foo%%%%bar\\%baz%", Some('\\')),
+            r#"^.*?foo.*?bar%baz.*?$"#
+        );
+        assert_eq!(
+            like_to_re_pattern("%foo%%%%bar*%baz%", Some('*')),
+            r#"^.*?foo.*?bar%baz.*?$"#
+        );
+        assert_eq!(like_to_re_pattern("_foo", Some('\\')), r#"^.foo$"#);
+        assert_eq!(like_to_re_pattern("foo_", Some('\\')), r#"^foo.$"#);
+        assert_eq!(like_to_re_pattern("foo_bar", Some('\\')), r#"^foo.bar$"#);
+        assert_eq!(like_to_re_pattern("foo__bar", Some('\\')), r#"^foo..bar$"#);
+        assert_eq!(
+            like_to_re_pattern("foo_.*?_bar", Some('\\')),
+            r#"^foo.\.\*\?.bar$"#
+        );
+    }
+
+    #[test]
+    fn like_match() {
+        let pat = like_to_re_pattern("foo_.*?_bar", Some('\\'));
+        let re = Regex::new(&pat).unwrap();
+
+        assert!(re.is_match("foos.*?qbar"));
+    }
+
+    #[test]
+    fn similar() {
+        assert_eq!(similar_to_re_pattern("(b|c)%", Some('\\')), r#"^(b|c).*?$"#);
+        assert_eq!(
+            similar_to_re_pattern("%(b|d)%", Some('\\')),
+            r#"^.*?(b|d).*?$"#
+        );
+    }
+
+    #[test]
+    fn similar_match() {
+        let pat = similar_to_re_pattern("(b|c)%", Some('\\'));
+        let re = Regex::new(&pat).unwrap();
+        assert!(!re.is_match("abc"));
+
+        let pat = similar_to_re_pattern("%(b|d)%", Some('\\'));
+        let re = Regex::new(&pat).unwrap();
+        assert!(!re.is_match("abc"));
```

Comment:
Btw, this assertion fails in the GH Actions build: https://github.com/partiql/partiql-lang-rust/actions/runs/3859256107/jobs/6578608692#step:7:456